src/uri/punycode.cr

```diff
@@ -0,0 +1,188 @@
+# `Punycode` provides an interface for IDNA encoding (RFC 5980),
+# which is defined in RFC 3493
+#
+# Implementation based on Mathias Bynens `punnycode.js` project
+# https://github.com/bestiejs/punycode.js/
+#
+# RFC 3492:
+# Method to use non-ascii characters as host name of URI
+# https://www.ietf.org/rfc/rfc3492.txt
+#
+# RFC 5980:
+# Internationalized Domain Names in Application
+# https://www.ietf.org/rfc/rfc5980.txt
+class URI
+  class Punycode
+    BASE         =  36
+    TMIN         =   1
+    TMAX         =  26
+    SKEW         =  38
+    DAMP         = 700
+    INITIAL_BIAS =  72
+    INITIAL_N    = 128
+
+    DELIMITER = '-'
+
+    BASE36 = "abcdefghijklmnopqrstuvwxyz0123456789"
+
+    private def self.adapt(delta, numpoints, firsttime)
+      delta /= firsttime ? DAMP : 2
+      delta += delta / numpoints
+      k = 0
+      while delta > ((BASE - TMIN) * TMAX) / 2
+        delta /= BASE - TMIN
+        k += BASE
+      end
+      k + (((BASE - TMIN + 1) * delta) / (delta + SKEW))
+    end
+
+    def self.encode(string : String)
+      encode string.chars
```

Comment:
It would be better to use a Char::Reader instead of allocating an array of chars. I don't think there is a real usecase for supplying a char array, is it?

Reply:
Sure.